fix: cost of poor quality report time filters not working

aaronmenezes opened the pull request:
#### Issue :
In the Cost of Poor Quality report, Time and date filter had no effect on the report.

#### Fix:
Update report query to fix this incorrect behavior.

aaronmenezes committed:
fix: cost of poor quality report time filters not working

ankush commented on erpnext/manufacturing/report/cost_of_poor_quality_report/cost_of_poor_quality_report.py:
```diff
@@ -20,30 +19,39 @@ def get_data(report_filters):
-		for row in job_cards:
-			row.operating_cost = flt(row.hour_rate) * (flt(row.total_time_in_mins) / 60.0)
-			data.append(row)
+		data = frappe.db.sql("""SELECT {fields} FROM `tabJob Card` tjc {filters}
```
Translate this to use query builder now 😅 

We are gearing up for Postgres support, so new raw queries won't be allowed from now on. 

https://frappeframework.com/docs/user/en/api/query-builder

aaronmenezes replied:
Okay. I will translate the query to use QB.

aaronmenezes committed:
chore:update cost of poor quality report to use query builder

aaronmenezes committed:
fix: linter warnings

aaronmenezes committed:
chore: updated report query

aaronmenezes committed:
Merge branch 'frappe:develop' into fix_cost_of_poor_quality_filters

aaronmenezes committed:
chore: added test filters

aaronmenezes committed:
fix : cleared linter warnings

ankush committed:
Merge branch 'develop' into fix_cost_of_poor_quality_filters

ankush commented:
report not usable right now on `develop` due to date parsing issue, probably introduced via https://github.com/frappe/frappe/pull/13504 

edit: fixed via https://github.com/frappe/frappe/pull/15454

aaronmenezes committed:
Merge branch 'frappe:develop' into fix_cost_of_poor_quality_filters

marination committed:
Merge branch 'develop' into fix_cost_of_poor_quality_filters

ankush commented on erpnext/manufacturing/report/cost_of_poor_quality_report/cost_of_poor_quality_report.py:
```diff
@@ -18,31 +17,45 @@ def get_data(report_filters):
+												))
+				)
+		query = append_filters(report_filters,operations,query,job_card)
+		data = query.run()
```
the previous function was returning a list of dicts. this is list of tuples. 🤔

ankush commented on erpnext/manufacturing/report/cost_of_poor_quality_report/cost_of_poor_quality_report.py:
```diff
@@ -18,31 +17,45 @@ def get_data(report_filters):
-				filters[field] = ('like', '% {} %'.format(report_filters.get(field)))
-
-	return filters
+				query = query.where(job_Card[field] ==report_filters.get(field))
```
```suggestion
				query = query.where(job_Card[field] == report_filters.get(field))
```

ankush commented on erpnext/manufacturing/report/cost_of_poor_quality_report/cost_of_poor_quality_report.py:
```diff
@@ -18,31 +17,45 @@ def get_data(report_filters):
 
-def get_filters(report_filters, operations):
-	filters = {"docstatus": 1, "operation": ("in", operations), "is_corrective_job_card": 1}
+def append_filters(report_filters, operations,query,job_Card):
```
```suggestion
def append_filters(report_filters, operations, query, job_Card):
```

ankush commented on erpnext/manufacturing/report/cost_of_poor_quality_report/cost_of_poor_quality_report.py:
```diff
@@ -18,31 +17,45 @@ def get_data(report_filters):
-def get_filters(report_filters, operations):
-	filters = {"docstatus": 1, "operation": ("in", operations), "is_corrective_job_card": 1}
+def append_filters(report_filters, operations,query,job_Card):
 	for field in ["name", "work_order", "operation", "workstation", "company", "serial_no", "batch_no", "production_item"]:
```
```suggestion
	for field in ("name", "work_order", "operation", "workstation",
					"company", "serial_no", "batch_no", "production_item"):
```

ankush commented on erpnext/manufacturing/report/cost_of_poor_quality_report/cost_of_poor_quality_report.py:
```diff
@@ -18,31 +17,45 @@ def get_data(report_filters):
+						.select(job_card_time_log.parent)
+						.where(job_card_time_log.parent == job_card.name)
+						.where(job_card_time_log.from_time >= report_filters.get('from_date',''))
+						.where(job_card_time_log.to_time <= report_filters.get('to_date',''))
```
if comparing an empty string like `x <= ''` nothing will show up I think? 🤔 

These time filters should be conditionally chained like other filters.

ankush commented on erpnext/manufacturing/report/cost_of_poor_quality_report/cost_of_poor_quality_report.py:
```diff
@@ -18,31 +17,45 @@ def get_data(report_filters):
+					job_card.serial_no, job_card.batch_no, job_card.workstation, job_card.total_time_in_mins, job_card.hour_rate,
+					operatingcost)
+					.where(job_card.docstatus==1)
+					.where(job_card.is_corrective_job_card==1)
```
- This condition should also be added in the subquery for faster filtering. 
- Sub-query should only be required when from/to time is specified.

ankush committed:
chore: formatting

ankush committed:
refactor: query generation

- optionally apply date filters
- join instead of expensive sub-query
- return as dictionary

ankush committed:
test: simplify test
